### PR TITLE
Fix isLatest calculation in CI and move tool to new CLI

### PR DESCRIPTION
**Pull request by tylerbutler**
Replaces PR #11537

Alternative approach to fixing the isLatest calculation in CI.

1. Refactored existing buildVersion command to a lib module and a main module, to make it easier to export the minimum stuff from build-tools.
2. Created a new command in the new CLI tools, `generate buildVersion`.
3. Updated CI to install the new build-cli package as part of the pipeline.
4. Use new CLI to run buildVersion in CI.

### CODE REVIEW

File: tools/pipelines/templates/include-set-package-version.yml [hunk 1]
```diff
@@ -9,21 +9,37 @@ parameters:
...
     workingDirectory: ${{ parameters.buildDirectory }}
     script: |
+      echo "$PATH"
       pushd "$(Build.SourcesDirectory)/build-tools"
       npm ci
-      npm run install:build-tools
       popd
-      npm link "$(Build.SourcesDirectory)/build-tools/packages/build-tools"
+
+      flub
```
Comment (tylerbutler) [inline tools/pipelines/templates/include-set-package-version.yml:42]:
```suggestion
      flub
```

**Comment by tylerbutler** [comment]
@curtisman @Josmithr This is a replacement for PR #11537 which you two previously reviewed. FYI in case you want to take another look.

File: build-tools/packages/build-cli/README.md [hunk 2]
```diff
@@ -209,6 +210,36 @@ DESCRIPTION
 
 _See code: [@oclif/plugin-commands](https://github.com/oclif/plugin-commands/blob/v2.2.0/src/commands/commands.ts)_
 
+## `flub generate buildVersion [FILE]`
+
+This command is used to compute the version number of Fluid packages. The release version number is based on what's in the lerna.json/package.json. The CI will supply the build number and branch to determine the prerelease suffix if it is not a tagged build
+
+```
```
Comment (Josmithr) [inline build-tools/packages/build-cli/README.md:217]: Any language we should specify here?

File: build-tools/packages/build-cli/src/commands/generate/buildVersion.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/* eslint-disable camelcase */
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/commands/generate/buildVersion.ts:6]: This is for `child_process` right? Would it be better to just import it as `ChildProcess` instead?

File: build-tools/packages/build-cli/src/commands/generate/buildVersion.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+import child_process from "child_process";
+import * as fs from "fs";
+import { getSimpleVersion, getIsLatest } from "@fluidframework/build-tools";
+import { Flags } from "@oclif/core";
+import { BaseCommand } from "../../base";
+
+export default class GenerateBuildVersionCommand extends BaseCommand<
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/commands/generate/buildVersion.ts:14]: Can we add a doc comment to this with a brief overview (or link to one that lives elsewhere)?

File: build-tools/packages/build-tools/src/buildVersion/buildVersionLib.ts [hunk 5]
```diff
@@ -0,0 +1,178 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/**
+ * This script is used by the build server to compute the version number of the packages.
+ * The release version number is based on what's in the lerna.json/package.json.
+ * The CI will supply the build number and branch to determine the prerelease suffix if it is not a tagged build
```
Comment (Josmithr) [inline build-tools/packages/build-tools/src/buildVersion/buildVersionLib.ts:9]:
```suggestion
 * The CI will supply the build number and branch to determine the prerelease suffix if it is not a tagged build.
```

File: build-tools/packages/build-tools/src/buildVersion/buildVersionLib.ts [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+ *      The computed version output to the console.
+ */
+
+import child_process from "child_process";
+import fs from "fs";
+import { detectVersionScheme, getLatestReleaseFromList, isInternalVersionScheme } from "@fluid-tools/version-tools";
+import * as semver from "semver";
+import { Logger } from "../common/logging";
+
+export function getFileVersion() {
```
Comment (Josmithr) [inline build-tools/packages/build-tools/src/buildVersion/buildVersionLib.ts:26]: Docs?

File: build-tools/packages/build-tools/src/buildVersion/buildVersionLib.ts [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    if (build_suffix) {
+        return `${release_version}-${build_suffix}`;
+    }
+
+    return release_version;
+}
+
+export function getSimpleVersion(file_version: string, arg_build_num: string, arg_release: boolean, patch: boolean) {
```
Comment (Josmithr) [inline build-tools/packages/build-tools/src/buildVersion/buildVersionLib.ts:84]: Docs?

File: build-tools/packages/build-tools/src/bumpVersion/gitRepo.ts [hunk 8]
```diff
@@ -170,9 +170,19 @@
...
         return await this.exec(`tag -l ${pattern}`, `get tags ${pattern}`);
     }
 
+    /**
+     * Get all tags matching a pattern.
+     *
+     * @param pattern pattern of tags to get
+     */
+    public async getAllTags(pattern?: string): Promise<string[]> {
+        const results = pattern === undefined ? await this.exec(`tag -l`, `get all tags`) : await this.exec(`tag -l "${pattern}"`, `get tags ${pattern}`);
```
[the remark this answers is not in the document]
Comment (Josmithr) [reply build-tools/packages/build-tools/src/bumpVersion/gitRepo.ts:182]: This warning appears in the new code. Not sure what are policy requirements are here, but figured I would note it,

File: build-tools/packages/version-tools/src/schemes.ts [hunk 9]
```diff
@@ -241,3 +253,12 @@ export function getLatestReleaseFromList(versionList: string[], allowPrereleases
 
     return latest;
 }
+
+export function getVersionFromTag(tag: string): string | undefined {
```
Comment (Josmithr) [inline build-tools/packages/version-tools/src/schemes.ts:257]: Docs?

File: tools/pipelines/templates/include-set-package-version.yml [hunk 10]
```diff
@@ -44,8 +60,10 @@ steps:
       echo TEST_BUILD=$TEST_BUILD
       echo VERSION_RELEASE=$VERSION_RELEASE
       echo VERSION_PATCH=$VERSION_PATCH
+      echo VERSION_INCLUDE_INTERNAL_VERSIONS=$VERSION_INCLUDE_INTERNAL_VERSIONS
 
-      fluid-build-version
+      flub --help
```
Comment (Josmithr) [inline tools/pipelines/templates/include-set-package-version.yml:65]: Is this intentional? Just to display some info about the script before running it?

**Review by Josmithr — APPROVED** [review]

File: tools/pipelines/templates/include-set-package-version.yml (hunk 10 above)
Comment (tylerbutler) [reply tools/pipelines/templates/include-set-package-version.yml:65]: Yeah. I'll add a comment.

File: build-tools/packages/build-cli/README.md (hunk 2 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/README.md:217]: This is auto-generated by oclif.

Comment (Josmithr) [reply build-tools/packages/build-cli/README.md:217]: Okay. I thought maybe we might be able to configure it to spit out `bash` or something for better syntax highlighting. But it's not a big deal.

**Review by Josmithr — APPROVED** [review]

**Pull request merged**